devdoc/channel.md

```diff
@@ -0,0 +1,297 @@
+
+ - **SRS_CHANNEL_43_103: [** `channel_pull` shall create a `THANDLE(ASYNC_OP)` by calling `async_op_create` with `cancel_channel_op` as `cancel`. **]**
+
+ - **SRS_CHANNEL_43_104: [** `channel_pull` shall store the `pull_callback` and `pull_context` in the `THANDLE(ASYNC_OP)`. **]**
+
+ - **SRS_CHANNEL_43_105: [** `channel_pull` shall insert the created `THANDLE(ASYNC_OP)` in the list of pending operations by calling `DList_InsertTailList`. **]**
+
+ - **SRS_CHANNEL_43_106: [** set the `state` to `CHANNEL_PULLED`. **]**
+
+ - **SRS_CHANNEL_43_107: [** set `*out_op_pull` to the created `THANDLE(ASYNC_OP)`. **]**
+
+**SRS_CHANNEL_43_108: [** If the `state` of the `channel` is `CHANNEL_PUSHED`: **]**
+
+ - **SRS_CHANNEL_43_109: [** `channel_pull` shall call `DList_RemoveHeadList` on the list of pending operations to obtain the `operation`. **]**
+
+ - **SRS_CHANNEL_43_110: [** If the list of pending operations is now empty, `channel_pull` shall set the `state` to `CHANNEL_EMPTY`. **]**
+
+ - **SRS_CHANNEL_43_111: [** `channel_pull` shall set the `state` of the obtained `operation` to `CHANNEL_OP_SCHEDULED`. **]**
+
+ - **SRS_CHANNEL_43_112: [** `channel_pull` shall store the `pull_callback` and `pull_context` in the obtained `operation`. **]**
+
+ - **SRS_CHANNEL_43_113: [** `channel_pull` shall call `threadpool_schedule_work` with `execute_callbacks` as `work_function` and the obtained `operation` as `work_function_context`. **]**
+
+ - **SRS_CHANNEL_43_114: [** `channel_pull` shall set `*out_op_pull` to the `THANDLE(ASYNC_OP)` of the obtained `operation`. **]**
+
+**SRS_CHANNEL_43_115: [** `channel_pull` shall call `srw_lock_release_exclusive`. **]**
+
+**SRS_CHANNEL_43_011: [** `channel_pull` shall succeeds and return `CHANNEL_RESULT_OK`. **]**
+
+**SRS_CHANNEL_43_023: [** If there are any failures, `channel_pull` shall fail and return `CHANNEL_RESULT_ERROR`. **]**
+
+
+### channel_push
+```c
+    MOCKABLE_FUNCTION(, CHANNEL_RESULT, channel_push, THANDLE(CHANNEL), channel, THANDLE(RC_PTR), data, PUSH_CALLBACK, push_callback, void*, push_context, THANDLE(ASYNC_OP)*, out_op_push);
+```
+
+`channel_push` notifies the channel that there is data available and registers the given `push_callback` to be called when the given `data` has been consumed.
+
+**SRS_CHANNEL_43_024: [** If `channel` is `NULL`, `channel_push` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_025: [** If `push_callback` is `NULL`, `channel_push` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_026: [** If `out_op_push` is `NULL`, `channel_push` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_116: [** `channel_push` shall call `srw_lock_acquire_exclusive`. **]**
+
+**SRS_CHANNEL_43_117: [** If the `state` of the `channel` is `CHANNEL_EMPTY` or `CHANNEL_PUSHED`: **]**
+
+ - **SRS_CHANNEL_43_118: [** `channel_push` shall call `sm_exec_begin`. **]**
```

Comment:
>sm_exec_begin [](http://example.com/codeflow?start=56&length=13)

the sm_exec_end that completes this sm_exec_begin is in line 297.

Meanwhile... a lot of things need to happen:
1) this function needs to exit (so it releases SRW_LOCK)
2) pull needs to be called (so it acquires SRW_LOCK) and pull must have advanced to line 

At this moment, we're inside `pull` function and pull advanced to line 203.

Note: at this moment we have sm_exec_begin and SRW_LOCK still taken.

Depending on the wits of the day of the threadpool, the SRW_LOCK might get released before or after the sm_exec_end is called.

Is that intended that sm's operation and the SRW_LOCK release can be completed in any order? Would that ever lead to a deadlock?

Reply:
I think that should be fine. Releasing the lock does not wait for sm_exec_end so the lock will always release.